05/app.js

```diff
@@ -0,0 +1,53 @@
+function Student(initialName, initialLastName) {
+    this.firstName = initialName;
+    this.lastName = initialLastName;
+    this.grades = {};
+}
+
+Student.prototype.addGrade = function (subject, grade) {
+    if (typeof this.grades[subject] === 'undefined') {
+        this.grades[subject] = [];
+    }
+
+    this.grades[subject].push(grade);
+}
+
+Student.prototype.getAvgGrade = function (subject) {
+    const subjectGrades = this.grades[subject];
+
+    // Przy wyliczeniu średniej ze wszystkich przedmiotów posiłkowałem się podobnym pytaniem na slacku
+    const grades = Array.from(Object.values(this.grades)).flat();
+
+
+
+    if (typeof subject !== 'undefined') {
+        const gradesSum = subjectGrades.reduce(function (previousValue, currentValue) {
+            return previousValue + currentValue;
+        }, 0)
+
+        const subjectAvg = gradesSum / subjectGrades.length;
+        return `Średnia z ${subject} to: ${subjectAvg.toFixed(2)}`;
+    } else {
+        const gradesSum = grades.reduce(function (previousValue, currentValue) {
+            return previousValue + currentValue;
+        }, 0)
+        const subjectsAvg = gradesSum / grades.length;
+        return `Średnia ze wszystkich przedmiotów to: ${subjectsAvg.toFixed(2)}`
+    }
```

Comment:
Zobacz, że kod tutaj mocno się powiela. Jedyna to różnica jakie dane wykorzystujemy do obliczenia średniej.
W takim wypadku idealnie nada się osobna funkcja, która liczy średnią, a dane (tablice) przekazujemy do niej przez parametr. Funkcja ta zwróci obliczoną średnią na podstawie danych przekazanych przez parametr. W pierwszym przypadku to `subjectGrades`, w drugim `grades`

Reply:
Super! Pozbyłem się kolejnych linijek kodu. 
Jeszcze mi się nasunęło jedno pytanie – w przypadku takich mini funkcji lepiej jest je implementować jako metoda do prototype obiektu, czy stworzyć zwykłą funkcję?